xapian-applications/omega/handler_libmwaw.cc

```diff
@@ -0,0 +1,264 @@
+
+    while (p != end) {
+	const char* start = p;
+	p = static_cast<const char*>(memchr(p, '\n', end - start));
+	const char* eol;
+	if (p)
+	    eol = p++;
+	else
+	    p = eol = end;
+	if ((end - start) > 5 && memcmp(start, "meta:", 5) == 0) {
+	    start += 5;
+	    switch (*start) {
+		case 'i': {
+		    if (author.empty())
+			PARSE_FIELD(start, eol, "initial-creator", author);
+		    break;
+		}
+		case 'k': {
+		    PARSE_FIELD(start, eol, "keyword", keywords);
+		    break;
+		}
+	    }
+	} else if ((end - start) > 3 && memcmp(start, "dc:", 3) == 0) {
+	    start += 3;
+	    switch (*start) {
+		case 'c': {
+		    if (!author.empty())
+			author.clear();
+		    PARSE_FIELD(start, eol, "creator", author);
+		    break;
+		}
+		case 's': {
+		    PARSE_FIELD(start, eol, "subject", keywords);
+		    break;
+		}
+		case 't': {
+		    PARSE_FIELD(start, eol, "title", title);
+		    break;
+		}
+	    }
+	} else if ((end - start) > 8 && memcmp(start, "dcterms:", 8) == 0) {
+	    start += 8;
+	    PARSE_FIELD(start, eol, "available", keywords);
+	}
+    }
+}
+
+static bool
+extract_word(string& dump,
+	     string& title,
+	     string& keywords,
+	     string& author,
+	     string& error,
+	     RVNGFileStream* input)
+{
+    RVNGString content_dump;
+    RVNGTextTextGenerator content(content_dump, false);
+    if (MWAWDocument::parse(input, &content) == MWAWDocument::MWAW_R_OK) {
+	dump = content_dump.cstr();
+    } else {
+	error = "Libmwaw Error: Failed to extract text";
+	return false;
+    }
+
+    RVNGString metadata_dump;
+    RVNGTextTextGenerator metadata(metadata_dump, true);
+    if (MWAWDocument::parse(input, &metadata) == MWAWDocument::MWAW_R_OK) {
+	const char* data = metadata_dump.cstr();
+	size_t len = metadata_dump.size();
+	parse_metadata(data, len, author, title, keywords);
+    } else {
+	error = "Libmwaw Error: Failed to extract metadata";
+	return false;
+    }
+    return true;
+}
+
+static bool
+extract_spreadsheet(string& dump,
+		    string& title,
+		    string& keywords,
+		    string& author,
+		    string& error,
+		    string& pages,
+		    RVNGFileStream* input)
+{
+    RVNGStringVector pages_metadata;
+    RVNGTextSpreadsheetGenerator metadata(pages_metadata, true);
+    MWAWDocument::Result result = MWAWDocument::parse(input, &metadata);
+    if (result != MWAWDocument::MWAW_R_OK) {
+	error = "Libmwaw Error: Failed to extract metadata";
+	return false;
+    } else {
+	string meta_temp;
+	for (unsigned i = 0; i < pages_metadata.size(); ++i) {
+	    meta_temp.append(pages_metadata[i].cstr());
+	}
+	const char* data = meta_temp.c_str();
+	size_t len = meta_temp.size();
+	parse_metadata(data, len, author, title, keywords);
+    }
+
+    RVNGStringVector pages_content;
+    RVNGTextSpreadsheetGenerator content(pages_content, false);
+    result = MWAWDocument::parse(input, &content);
+    if (result != MWAWDocument::MWAW_R_OK) {
+	error = "Libmwaw Error: Failed to extract text";
+	return false;
+    } else {
```

Comment:
Control flow can't leave the block above, because there's a `return` at the end.  In a situation like this, the code which follows doesn't need to be in an `else` block, and putting it in one tends to make the code less clear - the code which follows is conceptually part of the actual purpose of the function (vs the code in the `if` block above which is handling an error case.)

Reply:
Thanks for pointing this out. I have changed this.